ushahidi/platform
# 299/feat maintenance mode

This pull request makes the following changes:
- Adds fix for CORS header issue when Exceptions are raised before the Middleware stage of the Lumen Request cycle is reached
- Adds more specific error messages to explain failures on the API related to migration or deployment status
- Adds ability to place the API into Maintenance Mode and return this information to the client.

## Acceptance criteria

Test .env param for maintenance - on a local deployment
1. Set `MAINTENANCE_MODE=1` in your `.env` file
2. Load `api.ushahidi.test/api/v3/config/site` 
    - Should return a 503 with "SITE is down for maintenance"
3. Load client UI
    - See "SITE is down for maintenance" message.
4. Remove MAINTENANCE_MODE from .env
5. Load `api.ushahidi.test/api/v3/config/site` 
    - Should return 200 and config results

Test new deployment building messages:
1. Go to ushahididev.com/create (or equivalent)
2. Create a new deployment
3. Immediately try to load http://mydeployment.api.ushahididev.com/api/v3/config
    - Should return a 503 with "SITE is still being created"
4. Reload http://mydeployment.api.ushahididev.com/api/v3/config in 5 mins (ie. when it should be built)
    - Get 200 and site config data
5. Check cloud db deployments table and confirm `deployment_date` was set

Test maintenance mode:
1. Edit deployment entry in cloud.deployments and set status=maintenance
2. Try to load http://mydeployment.api.ushahididev.com/api/v3/config
    - Should return a 503 with "Site is down for maintenance"
3. Edit deployment entry in cloud.deployments and set status=migrating
4. Try to load http://mydeployment.api.ushahididev.com/api/v3/config
    - Should return a 503 with "Site is down for maintenance"
3. Edit deployment entry in cloud.deployments and set status=deployed
4. Try to load http://mydeployment.api.ushahididev.com/api/v3/config
    - Get 200 and site config data


Fixes [ushahidi/299](https://github.com/ushahidi/platform-cloud-interface/issues/299)

Ping @ushahidi/platform
